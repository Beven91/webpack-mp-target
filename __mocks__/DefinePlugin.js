module.exports = class DefinePlugin {
  constructor(options) {
    this.variables = options;
  }
  apply() {
  }
}