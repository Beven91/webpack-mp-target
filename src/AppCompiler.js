const DefinePlugin = require("webpack/lib/DefinePlugin");
const PlatformAdapter = require('./PlatformAdapter');
const platforms = require('./platforms');

module.exports = class MpAppCompiler {
  constructor(compiler, options) {
    this.compiler = compiler;
    this.options = {
      from: options.from || 'wx',
      target: options.target || 'wx',
    }
    // 原代码平台类型
    this.fromPlatform = platforms[this.options.from];
    // 编译的目标平台
    this.targetPlatform = platforms[this.options.target];
    // 校验参数
    if (this.validate()) {
      // 初始化
      this.initialize();
    }
  }

  // 获取全局变量名转换配置
  variables() {
    const defines = {};
    // 这里将来源的全局变量 同意替换成目标的全局变量 ，例如: wx ---> tt
    defines[this.fromPlatform.global] = this.targetPlatform.global;
    return defines;
  }

  initialize() {
    const compiler = this.compiler;
    compiler.plugin('emit', (compilation, callback) => {
      const adapter = new PlatformAdapter(compilation, this);
      adapter.transform();
      callback();
    });
    compiler.apply(
      // 将全局变量进行重命名
      new DefinePlugin(this.variables())
    )
  }

  validate() {
    const platformsKeys = Object.keys(platforms).join(',')
    if (!this.fromPlatform) {
      throw new Error(`不支持的原代码类型from:${this.options.from},\n仅支持以下类型\n:${platformsKeys}`);
    } else if (!this.targetPlatform) {
      throw new Error(`不支持的目标类型target:${this.options.target},\n仅支持以下类型\n:${platformsKeys}`);
    }
    return (this.fromPlatform !== this.targetPlatform);
  }
}