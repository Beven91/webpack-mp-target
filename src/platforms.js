module.exports = {
  'wx': {
    template: '.wxml',
    css: '.wxss',
    global: 'wx',
  },
  'alipay': {
    template: '.axml',
    css: '.axss',
    global: 'a'
  },
  'bytedance': {
    template: '.ttml',
    css: '.ttss',
    global: 'tt'
  }
}