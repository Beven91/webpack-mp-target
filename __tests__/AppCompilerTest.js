/**
 * @module AppCompiler
 * @name 测试 AppCompiler模块
 * @description
 */

import AppCompiler from '../src/AppCompiler';
import MockCompiler from 'webpack/lib/Compiler';
import DefinePlugin from 'webpack/lib/DefinePlugin';
import platforms from '../src/platforms';

describe('AppCompiler', () => {
  it('constructor', () => {
    const mockCompiler = new MockCompiler();
    const compiler = new AppCompiler(mockCompiler, { target: 'bytedance' })
    // 断言: 默认from为微信小程序
    expect(compiler.fromPlatform).toBe(platforms.wx);
    // 断言: 目标平台为 字节跳动小程序
    expect(compiler.targetPlatform).toBe(platforms.bytedance);
    // 断言: 当前变量值匹配镜像
    expect(compiler.variables()).toMatchSnapshot();

    // 创建指定类型的来源
    const compiler2 = new AppCompiler(mockCompiler, { from: 'alipay', target: 'wx' });
    // 断言: 默认from为微信小程序
    expect(compiler2.fromPlatform).toBe(platforms.alipay);
    // 断言: 目标平台为 字节跳动小程序
    expect(compiler2.targetPlatform).toBe(platforms.wx);
  });

  it('validate', () => {
    const mockCompiler = new MockCompiler();
    try {
      const compiler = new AppCompiler(mockCompiler, { from: 'xx' });
    } catch (ex) {
      expect(ex.message).toContain('不支持的原代码类型from');
    }
    try {
      new AppCompiler(mockCompiler, { target: 'xx' });
    } catch (ex) {
      expect(ex.message).toContain('不支持的目标类型target');
    }

    // 测试来源平台与目标平台一致，时不会进行任何处理
    const fn = jest.spyOn(AppCompiler.prototype, 'initialize')
    const compiler = new AppCompiler(mockCompiler, { from: 'wx', target: 'wx' });
    // 断言:在同一个平台下，不进行任何处理即 initialize不会被调用
    expect(fn).not.toHaveBeenCalled();

    // 取消监听
    fn.mockRestore();
  })

  it('initialize', () => {
    const mockCompiler = new MockCompiler();
    const spyInitialize = jest.spyOn(AppCompiler.prototype, 'initialize')
    const compiler = new AppCompiler(mockCompiler, { from: 'wx', target: 'alipay' });

    // 断言： initialize被调用
    expect(spyInitialize).toHaveBeenCalled();
    // 断言：使用了插件 DefinePlugin,且其配置，匹配镜像
    const definePlugin = mockCompiler._plugins.find((f) => f instanceof DefinePlugin);
    expect(definePlugin.variables).toMatchSnapshot();

    // 执行emit
    mockCompiler.emitAssets();
  })
});
