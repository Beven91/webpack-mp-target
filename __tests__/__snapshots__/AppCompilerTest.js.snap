// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AppCompiler constructor 1`] = `
Object {
  "wx": "tt",
}
`;

exports[`AppCompiler initialize 1`] = `
Object {
  "wx": "a",
}
`;
